videoplayer/src/darwin/videoplayer_darwin_viewcontroller.mm

```diff
@@ -166,57 +213,85 @@ -(void) observeValueForKeyPath:(NSString*)keyPath ofObject:(id)object change:(NS
         }
     }
-
-    // Error!
-    dmVideoPlayer::Command cmd;
-    memset(&cmd, 0, sizeof(cmd));
-    cmd.m_Type          = dmVideoPlayer::CMD_PREPARE_ERROR;
-    cmd.m_ID            = info->m_VideoId;
-    cmd.m_Callback      = info->m_Callback;
-    cmd.m_Width         = (int)info->m_Width;
-    cmd.m_Height        = (int)info->m_Height;
-    CommandQueue::Queue(&cmd);
 }
 
 - (void)PlayerItemDidReachEnd:(NSNotification *)notification { 
+    dmLogInfo("Videoplayer: PlayerItemDidReachEnd! video id: %d", m_SelectedVideoId);
+    
     SDarwinVideoInfo& info = m_Videos[m_SelectedVideoId];
-    m_SelectedVideoId = -1;
-    dmVideoPlayer::Command cmd;
-    memset(&cmd, 0, sizeof(cmd));
-    cmd.m_Type          = dmVideoPlayer::CMD_FINISHED;
-    cmd.m_ID            = info.m_VideoId;
-    cmd.m_Callback      = info.m_Callback;
-    cmd.m_Width         = (int)info.m_Width;
-    cmd.m_Height        = (int)info.m_Height;
-    CommandQueue::Queue(&cmd);
+    m_SelectedVideoId = INALID_VIDEO_ID;
+    m_ResumeOnForeground = false;
+    QueueVideoCommand(dmVideoPlayer::CMD_FINISHED, info);
+}
+
+-(void) AppEnteredBackground {
+    dmLogInfo("Videoplayer: AppEnteredBackground! video id: %d", m_SelectedVideoId);
+    if(m_SelectedVideoId != INALID_VIDEO_ID) {
+        dmLogInfo("Videoplayer: AppEnteredBackground: Pause video");
+
+        SDarwinVideoInfo& info = m_Videos[m_SelectedVideoId];
+        
+        [info.m_Player pause];
+        m_PauseTime = [info.m_Player currentTime];
+        m_ResumeOnForeground = true;
+    }
+}
+
+- (void) AppEnteredForeground {
+    dmLogInfo("Videoplayer: AppEnteredForeground! video id: %d", m_SelectedVideoId);
+    if((m_SelectedVideoId != INALID_VIDEO_ID) && m_ResumeOnForeground) {
+        dmLogInfo("Videoplayer: AppEnteredForeground: Resume video");
+
+        SDarwinVideoInfo& info = m_Videos[m_SelectedVideoId];
+
+        [info.m_PlayerLayer setPlayer:nil];
+        [info.m_Player replaceCurrentItemWithPlayerItem: nil];
+        [info.m_Player replaceCurrentItemWithPlayerItem: info.m_PlayerItem];
+        [info.m_PlayerLayer setPlayer:info.m_Player];
+        
+        [self ResumeFromPauseTime];
+        m_ResumeOnForeground = false;
+    }
+}
+
+-(void) ResumeFromPauseTime {
+    dmLogInfo("Videoplayer: ResumeFromPauseTime! video id: %d", m_SelectedVideoId);
+
+    if(m_SelectedVideoId != INALID_VIDEO_ID) {
+        SDarwinVideoInfo& info = m_Videos[m_SelectedVideoId];
+        AVPlayer* player = info.m_Player;
+        
+        if(player.status == AVPlayerStatusReadyToPlay && player.currentItem.status == AVPlayerItemStatusReadyToPlay) {
+            [player seekToTime:m_PauseTime toleranceBefore:kCMTimeZero toleranceAfter:kCMTimeZero completionHandler:^(BOOL finished) {
+                [player play];
+            }];
+        } else {
+            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
```

Comment:
"If the player isn't ready, wait a little bit and try resuming again"?

Reply:
Yes, as per stackoverflow wisdom.